### PR TITLE
add curl example for webhook invocations

### LINKED ISSUE #9543: [Documentation] - Add webhooks examples using curl
Would be great if there were some simple curl examples for triggering the build through hooks for each type: https://docs.openshift.org/latest/dev_guide/builds.html#webhook-triggers

CC: @mfojtik

### PR DESCRIPTION
**Pull request by bparees**
fixes https://github.com/openshift/origin/issues/9543

### CODE REVIEW

**Comment by bparees** [comment]
@csrwng ptal

**Comment by csrwng** [comment]
LGTM

**Comment by bparees** [comment]
@openshift/team-documentation ptal

**Comment by vikram-redhat** [comment]
@ahardin-rh  - can you please take a look today? 

@bparees - for origin AND enterprise branches?

**Comment by bparees** [comment]
@vikram-redhat yes but OSE is broken until this goes in: https://github.com/openshift/origin/pull/10031
(can't handle yaml payloads).

so maybe OSE 3.3 docs only?

**Comment by bparees** [comment]
that is, origin and OSE3.3 only.

File: dev_guide/builds.adoc [hunk 1]
```diff
@@ -1515,6 +1515,11 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
 trigger on your `*BuildConfig*` and trigger it via your Gogs server also.
 ====
 
+Given a file containing a valid json payload, you can manually trigger the webhook via `*curl*`:
+
+    $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST --data-binary @github_payload_file.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/github
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: Add ---- before and after the code snippet
(Per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference)

So: 
----
$ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST --data-binary @github_payload_file.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/github
----

File: dev_guide/builds.adoc [hunk 2]
```diff
@@ -1515,6 +1515,11 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
 trigger on your `*BuildConfig*` and trigger it via your Gogs server also.
 ====
 
+Given a file containing a valid json payload, you can manually trigger the webhook via `*curl*`:
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: s/json/JSON
only backticks around curl (no asterisks), as it is command

File: dev_guide/builds.adoc [hunk 3]
```diff
@@ -1515,6 +1515,11 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
...
+
+    $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST --data-binary @github_payload_file.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/github
+
+The `*-k*` argument is only necessary if your api server does not have a properly signed certificate.
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: s/api/API

File: dev_guide/builds.adoc [hunk 4]
```diff
@@ -1543,6 +1548,14 @@ webhook endpoint for your build:
 http://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/generic
 ----
 
+The caller must invoke the webhook as a `*POST*` operation.
+
+To invoke the webhook manually you can use `*curl*`:
+
+    $ curl -X POST -k https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/generic
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: As above, add ---- before and after the code snippet
(Per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference)

File: dev_guide/builds.adoc [hunk 5]
```diff
@@ -1543,6 +1548,14 @@ webhook endpoint for your build:
...
+
+The http verb must be set to `*POST*`.  The insecure `*-k*` flag is specified to ignore certificate validation.  This second flag is not necessary if your cluster has properly signed certificates.
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: s/http/HTTP
Only single spaces between sentences.

Comment (bparees) [inline dev_guide/builds.adoc]: pretty sure the formatter strips the spaces?

File: dev_guide/builds.adoc [hunk 6]
```diff
@@ -1575,6 +1588,12 @@ are ignored. Set the `*allowEnv*` field to `*true*` on the webhook definition to
 this behavior.
 ====
 
+To pass this payload using `*curl*`, define it in a file named `*payload_file.yaml*` and run:
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: only backticks around curl (no asterisks), as it is command
Change formatting of payload_file.yaml to:

     *_payload_file.yaml_*

File: dev_guide/builds.adoc [hunk 7]
```diff
@@ -1575,6 +1592,12 @@ are ignored. Set the `*allowEnv*` field to `*true*` on the webhook definition to
...
+
+    $ curl -H "Content-Type: application/yaml" --data-binary @payload_file.yaml -X POST -k https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/generic
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: Add --- before and after code snippet here too

File: dev_guide/builds.adoc [hunk 8]
```diff
@@ -1575,6 +1592,12 @@ are ignored. Set the `*allowEnv*` field to `*true*` on the webhook definition to
...
+
+The arguments are the same as the previous example with the addition of a header and a payload.  The `*-H*` argument sets the `*Content-Type*` header to `*application/yaml*` or `*application/json*` depending on your payload format.  The `*--data-binary*` argument is used to send a binary payload with newlines intact with the `*POST*` request.
```
Comment (ahardin-rh) [inline dev_guide/builds.adoc]: single spaces between all sentences

**Comment by ahardin-rh** [comment]
@bparees Just a few minor comments from me. Is there any reason why this should be not in Online or Dedicated 3.3? Thanks!

**Comment by bparees** [comment]
@ahardin-rh changes pushed.  as for putting it in online, well it won't work until online is updated to 3.3, again because of the yaml issue which my other PR will fix.

**Comment by ahardin-rh** [comment]
@bparees Thanks!

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../dev_guide/builds.adoc#dev-guide-builds[Builds]
|Added a `curl` example for webhook invocations within the xref:../dev_guide/builds.adoc#webhook-triggers[Webhook Triggers] section
%

**Pull request merged**